lms/djangoapps/mobile_api/users/views.py

```diff
@@ -502,3 +505,94 @@ class UserCourseEnrollmentsV4Pagination(DefaultPagination):
+        less than 30 days ago or has progressed in the course in the last 30 days.
+        Otherwise, the registration is considered inactive.
+
+    **Example Request**
+
+        GET /api/mobile/{api_version}/users/<user_name>/enrollments_status/
+
+    **Response Values**
+
+        If the request for information about the user's enrolments is successful, the
+        request returns an HTTP 200 "OK" response.
+
+        The HTTP 200 response has the following values.
+
+        * course_id (str): The course id associated with the user's enrollment.
+        * course_name (str): The course name associated with the user's enrollment.
+        * is_active (bool): User's course enrolment status.
+
+
+        The HTTP 200 response contains a list of dictionaries that contain info
+        about each user's enrolment status.
+
+    **Example Response**
+
+        ```json
+        [
+            {
+                "course_id": "course-v1:a+a+a",
+                "course_name": "a",
+                "is_active": true
+            },
+            {
+                "course_id": "course-v1:b+b+b",
+                "course_name": "b",
+                "is_active": true
+            },
+            {
+                "course_id": "course-v1:c+c+c",
+                "course_name": "c",
+                "is_active": false
+            },
+            ...
+        ]
+        ```
+    """
+    def get(self, request, *args, **kwargs) -> Response:
+        """
+        Gets user's enrollments status.
+        """
+        active_status_date = datetime.datetime.now(pytz.UTC) - datetime.timedelta(days=30)
+        user_enrollments = CourseEnrollment.objects.filter(
+            user__username=kwargs.get('username'),
+            is_active=True,
+        )
+        mobile_available = [
+            enrollment for enrollment in user_enrollments
+            if is_mobile_available_for_user(self.request.user, enrollment.course_overview)
+        ]
+        user_completions_last_month = BlockCompletion.objects.filter(
+            user__username=kwargs.get('username'),
+            created__gte=active_status_date
+        )
+        course_ids_where_user_has_completions = [
+            str(completion.block_key.course_key) for completion in user_completions_last_month
+        ]
+        enrollments_status = []
+
+        for user_enrollment in mobile_available:
+            course_id = str(user_enrollment.course_overview.id)
+            enrollments_status.append(
+                {
+                    'course_id': course_id,
+                    'course_name': user_enrollment.course_overview.display_name,
+                    'is_active': bool(
+                        course_id in course_ids_where_user_has_completions
+                        or user_enrollment.created > active_status_date
+                    )
+                }
+            )
+
+        return Response(enrollments_status)
```

Comment:
```suggestion
from django.db.models import Q

@mobile_view(is_user=True)
class UserEnrollmentsStatus(views.APIView):
    # ... docstring with more context ...

    def get(self, request, *args, **kwargs) -> Response:
        """Gets user's enrollments status."""
        enrollments = self._get_enrollments_with_progress(kwargs.get('username'))
        enrollments_status = self._build_enrollments_status(enrollments)
        return Response(enrollments_status)

    def _get_enrollments_with_progress(self, username):
        active_status_date = datetime.datetime.now(pytz.UTC) - datetime.timedelta(days=30)
        return CourseEnrollment.objects.filter(
            user__username=username,
            is_active=True,
            blockcompletion__created__gte=active_status_date
        ).select_related('course_overview')

    def _build_enrollments_status(self, enrollments):
        enrollments_status = []
        for enrollment in enrollments:
            if is_mobile_available_for_user(self.request.user, enrollment.course_overview):
                # ... (rest of the logic) ...
        return enrollments_status
```

This is suggestion from Gemini:

- I agree with them on the point to divide this big function into smaller functions.
- In terms of `Q` and combining queries I think it requires check if this will work as expected. So it is up to you if you think this is worth to change. It is more like FYI how it can be optimized.

Reply:
Yes, I think it's a good idea. I'm going to divide this big function into smaller functions